torch/distributed/_tensor/ops/math_ops.py

```diff
@@ -227,6 +232,57 @@ def var_reduction_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
     )
 
 
+@dataclass(frozen=True)
+class _NormPartial(_Partial):
+    norm_type: Union[int, float, str] = 2
+
+    def _partition_value(self):
+        raise NotImplementedError()
+
+    def _reduce_shard_value(self):
+        raise NotImplementedError()
+
+    def _reduce_value(
+        self, tensor: torch.Tensor, mesh: DeviceMesh, mesh_dim: int
+    ) -> torch.Tensor:
+        if self.norm_type in (float("inf"), "inf"):
+            return funcol.all_reduce(tensor, reduceOp="max", group=(mesh, mesh_dim))
+        elif self.norm_type in (float("-inf"), "-inf"):
+            return funcol.all_reduce(tensor, reduceOp="min", group=(mesh, mesh_dim))
+        elif self.norm_type == 0:
+            raise NotImplementedError()
+        elif isinstance(self.norm_type, (int, float)):
+            reduced_tensor = funcol.all_reduce(
+                tensor**self.norm_type, reduceOp="sum", group=(mesh, mesh_dim)
```

Comment:
hmmm wondering if this `tensor**self.norm_type` is needed? when doing local tensor computation in DTensor, it's calling `aten.linalg_vector_norm.default` on the local shards, and for each shard it would do the `**self.norm_type` inside the operator itself, so each shard should already been "scaled"

Reply:
I think that sum alone would not be the right reduction op. Let me try to write out an example.

Suppose we have two ranks, 0 and 1, and suppose we have two gradients with shapes `(4,)` and `(4,)`. Label the 1st shape `(4,)` gradient as $[g_1, g_2, g_3, g_4]$ and the 2nd shape `(4,)` gradient as $[g_5, g_6, g_7, g_8]$.
- Rank 0 has $[g_1, g_2]$, $[g_5, g_6]$.
- Rank 1 has $[g_3, g_4]$, $[g_7, g_8]$.

After the 1st vector norm and stack:
- Rank 0 has $[\sqrt{g_1^2 + g_2^2}, \sqrt{g_5^2 + g_6^2}]$.
- Rank 1 has $[\sqrt{g_3^2 + g_4^2}, \sqrt{g_7^2 + g_8^2}]$.

After the 2nd vector norm:
- Rank 0 has $[\sqrt{g_1^2 + g_2^2 + g_5^2 + g_6^2}]$.
- Rank 1 has $[\sqrt{g_3^2 + g_4^2 + g_7^2 + g_8^2}]$.

We want each rank to have $[\sqrt{\sum_{i \in [8]} g_i^2}]$. Therefore, we see that the reduction op needed is $f(x, y) = \sqrt{x^2 + y^2}$, not $f(x, y) = x + y$.

Reply:
I see, yeah this make sense! Do you mind putting the math formulas as a comment in the _NormPartial?